fabric-client-tags-v1/src/client/java/net/fabricmc/fabric/api/tag/client/v1/ClientTags.java

```diff
@@ -0,0 +1,161 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.tag.client.v1;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import it.unimi.dsi.fastutil.objects.Reference2ObjectOpenHashMap;
+
+import net.minecraft.client.MinecraftClient;
+import net.minecraft.tag.TagKey;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.registry.Registry;
+import net.minecraft.util.registry.RegistryEntry;
+import net.minecraft.util.registry.RegistryKey;
+
+import net.fabricmc.api.EnvType;
+import net.fabricmc.api.Environment;
+import net.fabricmc.fabric.impl.tag.client.DataLoader;
+
+/**
+ * Tags are loaded by the server, either the internal server in singleplayer or the connected server and
+ * synced to the client. This can be a pain point for interoperability, as a tag that does not exist on the server
+ * because it is part of a mod only present on the client will no longer be available to the client that may wish to
+ * query it.
+ *
+ * <p>Client Tags resolve that issue by lazily reading the tag json files within the mods on the side of the caller,
+ * directly, allowing for mods to query tags such as {@link net.fabricmc.fabric.api.tag.convention.v1.ConventionalBlockTags}
+ * even when connected to a vanilla server.
+ */
+@Environment(EnvType.CLIENT)
+public final class ClientTags {
+	private static final Map<TagKey<?>, Set<Identifier>> LOCAL_TAG_CACHE =
+			Collections.synchronizedMap(new Reference2ObjectOpenHashMap<>());
+	private static final DataLoader LOADER = new DataLoader();
+
+	private ClientTags() {
+	}
+
+	/**
+	 * Loads a tag into the cache, recursively loading any contained tags along with it.
+	 *
+	 * @param tagKey the {@code TagKey} to load
+	 * @return a set of {@code Identifier}s this tag contains
+	 */
+	public static Set<Identifier> getOrCreateLocalTag(TagKey<?> tagKey) {
+		return LOCAL_TAG_CACHE.computeIfAbsent(tagKey, LOADER::loadTag);
+	}
+
+	/**
+	 * Checks if an entry is in a tag.
+	 *
+	 * <p>If the synced tag does exist, it is queried. If it does not exist,
+	 * the tag populated from the available mods is checked.
+	 *
+	 * @param tagKey the {@code TagKey} to being checked
+	 * @param entry  the entry to check
+	 * @return if the entry is in the given tag
+	 */
+	@SuppressWarnings("unchecked")
+	public static <T> boolean isInWithLocalFallback(TagKey<T> tagKey, T entry) {
+		Optional<? extends Registry<?>> maybeRegistry = getRegistry(tagKey);
+
+		if (maybeRegistry.isPresent()) {
+			if (tagKey.isOf(maybeRegistry.get().getKey())) {
+				Registry<T> registry = (Registry<T>) maybeRegistry.get();
+
+				Optional<RegistryKey<T>> maybeKey = registry.getKey(entry);
+
+				// Check synced tag
+				if (registry.containsTag(tagKey)) {
+					return maybeKey.filter(registryKey -> registry.entryOf(registryKey).isIn(tagKey))
```

Comment:
I'm not a fan of the optional and filter stuff. This is not performance critical but the repeated allocations (including a capturing lambda) are quite gross imo. Up to you, don't consider this merge-blocking.

Reply:
I find it easier to read, so I'll leave it unless it actually causes an issue.